enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,1051 @@
+
+Both rollback and downgrade maybe look similar in terms of version change,
+the difference is that for rollback a matching backups exists,
+whereas for downgrade a data migration would have to be performed
+
+Decision to perform or refuse a data migration to schema compatible with newly loaded 
+MicroShift version will be based on following facts:
+- version persisted in MicroShift's data dir (version that created/successfully ran using the data),
+  also referred to as (version) metadata
+- version of currently installed MicroShift binary
+- embedded in MicroShift binary list of blocked "from" versions
+
+A general flow will have following form:
+1. If persisted version is missing, assume 4.13.
+1. If version of `microshift` binary is older than version in metadata, **refuse to start MicroShift**.
+1. If persisted version is on a list of blocked version migrations, **refuse to start MicroShift**.
+1. If binary is the same version as persisted in metadata, **no need for a data migration**.
+1. Otherwise attempt to migrate the data.
+
+
+### Open Questions [optional]
+
+#### Should 4.13 -> 4.14 be supported?
+- Even if we handle "no metadata, no backup, existing data, no next-boot-action" and make
+  a backup, upon rollback 4.13 won't be able to restore it, so it would try to use 4.14's
+  data.
+  - Is manual intervention acceptable (admin manually copying backup to data)?
+  - We most likely don't want to implement special case for that (`pre-run` would save 
+    that persist it was 4.13 previously and red script would restore before shutting down)
+
+#### If green/red info is not persisted, and it's not a first boot (/var/lib/microshift exists), what should we do?
+- Compare versions (metadata vs binary)
+  - Same - start cluster
+  - Different - means that both binaries implement the enhancement,
+    so this could be considered a bug.
+  - Metadata is also missing
+    - Assume data was created by 4.13, do a backup and follow data migration rules
+
+#### When should version metadata file be created
+- By the green script
+  - Existence of file would indicate that data belongs to healthy deployment
+    - What upon reboot? Should file be cleared to be later recreated by green script?
+      - Blind assumption `.version exists == healthy data` might be too implicit and lead
+        to wrong conclusions
+- On start of `microshift run` or shortly after with addition from green and red scripts
+  - `{ "version": "X.Y.Z", "status": "unknown"}` would be created at first
+  - Green script would change `.status` to `healthy`
+  - Red script would change `.status` to `unhealthy`
+  - Pros:
+    - Explicit
+  - Cons:
+    - More complicated schema - weak con
+
+#### How should `microshift pre-run` be executed?
+- `microshift.service` - `ExecStartPre`
+  - No need to add new systemd service files.
+  - It will run on each `systemctl restart microshift` which is not desirable (will it run when systemd restarts MicroShift?)
```

Comment:
Can you elaborate here on why it is not desirable?

I'm also interested in some detail about why `pre-run` is not just part of `run`. Why do we want separate commands?

Reply:
I was trying to ensure we run this only once, on boot, before MicroShift. Otherwise:
- should it run when admin issues `systemctl restart microshift`? In this case, we definately need to smartly handle missing pre-run-action
- should it run when MicroShift fails to be ready in 1 minute and systemd restarts the service and reruns the pre-run?

As for separating pre-run from run: to have separate distinct stages, if the pre-run.svc failed, microshift.svc won't even start, so it might be easier to track down the problem. If they're together, systemd will restart the microshift svc (indefinitely?) . 
We can make it work, but this might impact some assumptions we had about actions running on boot.
First is deletion of `pre-run-actions` after doing the work - it would have to persist for the whole boot, risking going stale (current cases where file would be missing might become invalid because it'll be an old file), also: file is created as an effect of green script, which will run about 2-3 minutes after boot (depending how long microshift healthcheck takes), so file right after boot and not so long after might be different.

Our discussion under code PR about dir structure and resemblance to `/ostree` made me wonder if we could think differently about backups and restores - `live/` would be a symlink to `deploy/id/`:
```
/var/lib/microshift/
   live/ -> deployments/rhel-123.0/
   deployments/
         rhel-123.0/
```
On boot, microshift would check if `live` points to data dir matching current deployment id - if not, then it'd create a new dir with reflinks to what `live` points to, switch `live` symlink.
One major change I see straight away is the approach of "retrying data migration" after failed 1st boot of new deployment (which now that I think about it is kind of a deviation from the rule, isn't it? would you expect data restore and migration on each red boot of new deployment from other programs? if we fail data migration in at least 95% cases it's probably a bug).
Is it too late to start considering it? :D

Reply:
Using a symlink for `live/` does seem appealing, but we will need to go back through all of our analysis to consider what impact that will have and if we need other changes. For one thing, there would no longer be "backups", just "old copies" of data. The semantics are a little different, so some of the actions we need to take might have to change.

You've expressed concerns about running migrations or other upgrade steps multiple times. I've been considering those as needing to be idempotent, so repeating them wouldn't cause any issues. Maybe that's not a safe assumption, and that's why you're worried about it?

Reply:
If we use data directories tied to ostree deployment IDs, what do we do on regular RPM systems without ostree? Is `live/` just a regular directory then?

Reply:
The gist you mention in https://github.com/openshift/enhancements/pull/1387#discussion_r1180297017 works through some of the changes I was worried about.

Is there ever a need to back up the data for the current ostree deployment before starting MicroShift? Under the old approach we would have retained the backup of the data for commit 1 and so could restore from it again if greenboot failed after a rollback. With the symlink approach, we would swap the live link but not take a backup, so we might lose data if something goes wrong and the rollback fails for some reason.

Reply:
> Is there ever a need to back up the data for the current ostree deployment before starting MicroShift? 

This what happens currently on reboots if previous boot was green, right? Or maybe "what if" is missing in the beginning and this concern about symlink approach?

> Under the old approach we would have retained the backup of the data for commit 1 and so could restore from it again if greenboot failed after a rollback. With the symlink approach, we would swap the live link but not take a backup, so we might lose data if something goes wrong and the rollback fails for some reason.

Yep, you're right. Let's keep direction straight

Reply:
We should capture this conclusion about why we want backups, rather than changing `live` to point to a different location, in the body of the text.